### PR TITLE
Remove DesugaringKind::Replace.

### PR DESCRIPTION
**Pull request by cjgillot**
A simple boolean flag is enough.

### CODE REVIEW

File: tests/ui/borrowck/issue-45199.rs [hunk 1]
```diff
@@ -14,14 +13,12 @@ fn test_call() {
...
     b = Box::new(2);        //~ ERROR cannot assign twice to immutable variable `b`
                             //~| NOTE cannot assign twice to immutable
-                            //~| NOTE in this expansion of desugaring of drop and replace
 }
 
 fn test_args(b: Box<i32>) {  //~ HELP consider making this binding mutable
                                 //~| SUGGESTION mut b
     b = Box::new(2);            //~ ERROR cannot assign to immutable argument `b`
                                 //~| NOTE cannot assign to immutable argument
-                                //~| NOTE in this expansion of desugaring of drop and replace
```
Comment (WaffleLapkin) [inline tests/ui/borrowck/issue-45199.rs:24]: Why  aren't those notes in the `.stderr` file? Like... I understand why they should not be there *anymore* but the diff of `.stderr` doesn't contain them at all?...

**Review by WaffleLapkin — APPROVED** [review]
The code changes look good to me. r=me with `.stderr` mystery resolved I guess

File: tests/ui/borrowck/issue-45199.rs (hunk 1 above)
Comment (cjgillot) [reply tests/ui/borrowck/issue-45199.rs:24]: This happens with all the macro backtrace "in this expansion" messages. This is probably deliberate, but I can't track where in code this comes from.

**Pull request merged**